### PR TITLE
Improve reliability of bots

### PR DESCRIPTION
**Pull request by MatthewWilkes**
## Description
This fixes a problem in serialising new participants that could cause high performance bots to get stuck in the waiting room. It also tidies some code around the questionnaire that was nonfunctional.

## Motivation and Context
This is part of a series on making bots work better.

## How Has This Been Tested?
This has been tested locally, @mmosleh is testing to see if it fixes https://github.com/Dallinger/Griduniverse/issues/196

### CODE REVIEW

File: dallinger/bots.py [hunk 1]
```diff
@@ -231,7 +232,12 @@ def sign_up(self):
                     hash=uuid.uuid4().hex
                 )
             )
-            result = requests.post(url)
+            try:
+                result = requests.post(url)
+            except ConnectionError:
```
Comment (alecpm) [inline dallinger/bots.py]: I'd suggest maybe using `result.raise_for_status()` here to combine the two calls to `self.stochastic_sleep()` into one exception handler with `requests.exceptions.RequestException`. That wouldn't handle the `json()[' status'] == 'error'` case though, but that case could do an explicit `raise requests.exceptions.HTTPError('JSON error status returned: {}'.format(result.json()))`. Then maybe add a `self.log("Bot request exception, retrying: {}".format(e))` in the exception handler to ensure we get information about these failures.

**Comment by vlall** [comment]
Needs a rebase

File: dallinger/bots.py (hunk 1 above)
Comment (MatthewWilkes) [reply dallinger/bots.py]: I feel a bit off about raising an explicit HTTP error here considering we're not putting it in the status. I think I'd rather have the 2 line duplication than the unusual raise. I think you're absolutely right about the `raise_for_status()` though, so I'm going to incorporate that.

**Comment by MatthewWilkes** [comment]
I've rebased this, addressed some (but not all) comments from @alecpm and merged some of the cleanup into the existing commits. That's in #1435

**Pull request closed without merging**